docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
> Flat namespaces allow you to easily deduplicate blobs shared between “repositories”.

The signatures are a few hundred bytes large; deduplicating them is fairly likely to cost more storage and slow things down than not doing anything. And of course the underlying storage is perfectly free to deduplicate anyway, by keeping a flat storage somewhere and replacing the $host/$repo@$digest directories with symlinks to the flat storage (or not even having physical directories and doing this all in server-side code). But a flat layout would make it impossible to provide privacy and access controls on the namespaces.

> They also make access more straightforward (because you only need to know the digest, and don't need to also know one repository with which the blob is associated).

That could only be useful if one wanted to _only_ wanted to know signatures for a digest, but I can’t see what good that would do. To do pretty much anything, the user needs the manifest as well as the signatures, and the manifest can only be optained from a repository; hence every user of a manifest always knows what a repository to use.

> Where privacy is a concern, a more reliable solution is for the publishers to encrypt the sensitive blob.

That’s pretty impractical for a cloud image distribution mechanism. Suddenly every client wanting to use these images would need to be set up private keys, _and they would have to be shared between all of the clients_, which makes them not private.

Ordinarily, the privacy would be provided by using plain HTTPS over the wire, which protects the network and allows authentication of the users using bearer tokens and the like. As for privacy against the registry…

> That way, you aren't leaking the sensitive data to the registry admins, and the registry admins can't leak the sensitive data to third parties (even accidentally).

… the registry admins necessarily know the existence of names of images; that’s how the clients ask for the images. If there is any privacy for cloud images at all, it is not against the registry admins; it’s between various groups using a (multi-tenant) registry.

It is _not_ an intention of this sigstore mechanism to add a new kind of privacy to the information whether a cloud image exists on a registry; but it would be nice for the sigstore mechanism not to make things worse WRT privacy.

> For content that is so sensitive that you aren't even comfortable publishing an encrypted form, you can set urls, serve the blob behind whatever access-control you feel is appropriate, 

_Yes, exactly_. And namespacing the signature storage using hosts and repo names makes access control much easier, because it can be namespaced per-host or per-repo, instead of setting up and constantly updating ACLs for every single image digest or possibly every single signature.